rectools/visuals/metrics_app.py

```diff
@@ -0,0 +1,212 @@
+                    mode="markers",
+                    name=model,
+                    hovertemplate=template,
+                    showlegend=self.show_legend,
+                )
+            fig_widget.layout.xaxis.title = metric_x.value
+            fig_widget.layout.yaxis.title = metric_y.value
+        self.fig = go.Figure(fig_widget.data)
+
+    def _update_chart(
+        self,
+        fig_widget: go.FigureWidget,
+        metric_x: widgets.Dropdown,
+        metric_y: widgets.Dropdown,
+        use_avg: widgets.Checkbox,
+        fold_i: widgets.Dropdown,
+    ) -> None:
+        data = self._make_chart_data_avg() if use_avg.value else self._make_chart_data(fold_i.value)
+        with fig_widget.batch_update():
+            existing_traces = {trace.name: trace for trace in fig_widget.data}
+            for model in self.model_names:
+                df = data[data[Columns.Model] == model]
+                trace = existing_traces.get(model)
+                template = f"<b>{model}</b><br>{metric_x.value}: %{{x}}<br>{metric_y.value}: %{{y}}<extra></extra>"
+                if trace is not None:
+                    trace.x = df[metric_x.value]
+                    trace.y = df[metric_y.value]
+                    trace.hovertemplate = template
+            fig_widget.layout.xaxis.title = metric_x.value
+            fig_widget.layout.yaxis.title = metric_y.value
+        self.fig = go.Figure(fig_widget.data)
+
+    def _update_fold_visibility(self, use_avg: widgets.Checkbox, fold_i: widgets.Dropdown) -> None:
+        fold_i.layout.visibility = "hidden" if use_avg.value else "visible"
+
+    def display(self) -> None:
+        """Display MetricsApp widget"""
+        metric_x = widgets.Dropdown(description="Metric X:", value=self.metric_names[0], options=self.metric_names)
+        metric_y = widgets.Dropdown(description="Metric Y:", value=self.metric_names[-1], options=self.metric_names)
+        use_avg = widgets.Checkbox(description="Avg folds", value=False)
+        fold_i = widgets.Dropdown(description="Fold number:", value=self.fold_ids[0], options=self.fold_ids)
+
+        layout_params = {
+            "width": WIDGET_WIDTH,
+            "height": WIDGET_HEIGHT,
+            "margin": {"t": TOP_CHART_MARGIN},
+            "legend_title": LEGEND_TITLE,
+        }
+        layout_params.update(self.layout_kwargs)
+        fig_widget = go.FigureWidget(layout=layout_params)
+
+        metric_x.observe(lambda upd: self._update_chart(fig_widget, metric_x, metric_y, use_avg, fold_i), "value")
+        metric_y.observe(lambda upd: self._update_chart(fig_widget, metric_x, metric_y, use_avg, fold_i), "value")
+        use_avg.observe(lambda upd: self._update_chart(fig_widget, metric_x, metric_y, use_avg, fold_i), "value")
+        use_avg.observe(lambda upd: self._update_fold_visibility(use_avg, fold_i), "value")
+        fold_i.observe(lambda upd: self._update_chart(fig_widget, metric_x, metric_y, use_avg, fold_i), "value")
```

Comment:
My prev. comment about duplicating lambdas still makes sense. Let's please move it to a separate function

Reply:
Unfortunately I can't find clean solution for this idea. We need to observe every widget and need to pass update methods through lambda function. 
We could significantly improve this part if we would keep all widgets as class variables, which I try to avoid to reduce coupling between display and data preparation logic.

Reply:
You can write a small function just before metric_x.observe like 
```
def update(event: ...) -> None:
    self.update_...
```
and use it instead of lambda

Reply:
Done